src/libraries/Microsoft.Extensions.Options.DataAnnotations/ref/Microsoft.Extensions.Options.DataAnnotations.cs

```diff
@@ -8,13 +8,15 @@ namespace Microsoft.Extensions.DependencyInjection
 {
     public static partial class OptionsBuilderDataAnnotationsExtensions
     {
-        public static Microsoft.Extensions.Options.OptionsBuilder<TOptions> ValidateDataAnnotations<TOptions>(this Microsoft.Extensions.Options.OptionsBuilder<TOptions> optionsBuilder) where TOptions : class { throw null; }
+        [System.Diagnostics.CodeAnalysis.RequiresUnreferencedCodeAttribute("Uses DataAnnotationValidateOptions which is unsafe given that the options type passed in when calling Validate cannot be statically analyzed so its members may be trimmed.")]
+        public static Microsoft.Extensions.Options.OptionsBuilder<TOptions> ValidateDataAnnotations<[System.Diagnostics.CodeAnalysis.DynamicallyAccessedMembers(System.Diagnostics.CodeAnalysis.DynamicallyAccessedMemberTypes.PublicProperties | System.Diagnostics.CodeAnalysis.DynamicallyAccessedMemberTypes.NonPublicProperties)] TOptions>(this Microsoft.Extensions.Options.OptionsBuilder<TOptions> optionsBuilder) where TOptions : class { throw null; }
     }
 }
 namespace Microsoft.Extensions.Options
 {
-    public partial class DataAnnotationValidateOptions<TOptions> : Microsoft.Extensions.Options.IValidateOptions<TOptions> where TOptions : class
+    public partial class DataAnnotationValidateOptions<[System.Diagnostics.CodeAnalysis.DynamicallyAccessedMembers(System.Diagnostics.CodeAnalysis.DynamicallyAccessedMemberTypes.PublicProperties | System.Diagnostics.CodeAnalysis.DynamicallyAccessedMemberTypes.NonPublicProperties)] TOptions> : Microsoft.Extensions.Options.IValidateOptions<TOptions> where TOptions : class
```

Comment:
I don't think this is correct because `IValidationOptions<T>` isn't annotated with the same `DAM` members.

@davidfowl @vitek-karas - this is the exact trap with DependencyInjection + trimming.

You think this pattern is safe because the linker doesn't warn, when in fact it is covered up by the warning inside DependencyInjection itself with the `MakeGenericType` call.

However, we may still be "safe" here because the ctor is annotated as RUC, and the only place we create one of these (in the above `ValidateDataAnnotations` API) is RUC as well.

thoughts?

Reply:
We would need a new linker feature to improve this - and to be honest I don't even know what that could look like right now.

I think the `ValidateDataAnnotations` is correct (since it's RUC), but the .ctor RUC itself doesn't help if I would call `services.AddSingleton<IValidateOptions<>,DataAnnotationValidateOptions<>>()`. That method lacks annotations currently - it will need to require .ctor on the implementation type - once that is done, even that should work since that should report the RUC on the .ctor.

Reply:
You're right - I was stupid enough to trust ILSpy to resolve the right dependency - and it resolved 3.1 version ... My bad.
In that case this should work (you would get a RUC warning when calling AddSingleton).